File: codewars/Ruby/7KYU/Alternate_capitalization.rb
=begin
Given a string, capitalize the letters that occupy even indexes and odd indexes separately, and return as shown below. Index 0 will be considered even.
For example, capitalize("abcdef") = ['AbCdEf', 'aBcDeF']. See test cases for more examples.
The input will be a lowercase string with no spaces.
Good luck!
=end

# My Solution
def capitalize(s)
  el = s.split('').map.with_index {|x,i| i % 2 === 0 ? x.upcase : x}.join('')
  [el, el.swapcase]
end

# Better Solution
def capitalize(s)
  result = s.gsub /.{1,2}/, &:capitalize
  [result, result.swapcase]
end

# Another Solution
def capitalize(s)
  result = s.split("").each_with_index.map {|x,i| i%2 == 0 ? x : x.upcase  }.join
  return [result.swapcase, result]
end

# Another Solution
def capitalize(s)
  res = []
  res << s.chars.map.with_index { |c, i| i % 2 == 0 ? c.upcase : c }.join
  res << res.first.swapcase
end

# Another Solution
def capitalize(str)
  f = ->str { str.scan(/\w{1,2}/).map(&:capitalize).join }
  [f.(str), str[0] + f.(str[1..-1])]
end

File: codewars/JavaScript/7KYU/Alternate_capitalization.js
/*
Given a string, capitalize the letters that occupy even indexes and odd indexes separately, and return as shown below. Index 0 will be considered even.
For example, capitalize("abcdef") = ['AbCdEf', 'aBcDeF']. See test cases for more examples.
The input will be a lowercase string with no spaces.
*/

// My Solution
const capitalize = s => s.split('').reduce((acc, rec, i) => {
  return [
    acc[0] + (i % 2 ? rec.toLowerCase() : rec.toUpperCase()),
    acc[1] + (i % 2 ? rec.toUpperCase() : rec.toLowerCase())
  ]
},['',''])

// Better Solution
function capitalize(s){
  const odd = s.split("").map((l, i) => i % 2 !== 0 ? l.toUpperCase() : l).join("");
  const even = s.split("").map((l, i) => i % 2 === 0 ? l.toUpperCase() : l).join("");
  return [even, odd];
};

// Another Solution
function capitalize(s){
  return [0,1].map(r=>[...s].map((c,i)=>i%2===r?c.toUpperCase():c).join(''));
};

// Another Solution
function capitalize(s){
  return [[...s].map((x,i) => i % 2 == 0 ? x.toUpperCase() : x).join(''),
          [...s].map((x,i) => i % 2 != 0 ? x.toUpperCase() : x).join('')]
}

// Another Solution
const cap = (str, isEven) => str.split('').map( (a, i) => isEven && i % 2 === 0 || !isEven && i % 2 === 1? a.toUpperCase(): a.toLowerCase()).join('');

function capitalize(s){
  return [cap(s, true), cap(s, false)]
};
